### ISSUE TITLE
Troubles reading DM4 4-D stack

### ISSUE THREAD

**Issue by woozey** [hyperspy/hyperspy#2031]
Got the following error trying to load large DM4 4-D stack file:

```python
ss = hs.load('D:\\Vadim\\2018-08-02_STEM-d_C_fiber_Fengshan\\bin2\\02_0V.dm4')

---------------------------------------------------------------------------
DM3TagIDError                             Traceback (most recent call last)
<ipython-input-20-bfa902970b32> in <module>()
----> 1 ss = hs.load('D:\\Vadim\\2018-08-02_STEM-d_C_fiber_Fengshan\\bin2\\02_0V.dm4')

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, lazy, **kwds)
    241             objects = [load_single_file(filename, lazy=lazy,
    242                                         **kwds)
--> 243                        for filename in filenames]
    244 
    245         if len(objects) == 1:

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in <listcomp>(.0)
    241             objects = [load_single_file(filename, lazy=lazy,
    242                                         **kwds)
--> 243                        for filename in filenames]
    244 
    245         if len(objects) == 1:

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load_single_file(filename, signal_type, **kwds)
    284                                 reader=reader,
    285                                 signal_type=signal_type,
--> 286                                 **kwds)
    287 
    288 

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load_with_reader(filename, reader, signal_type, **kwds)
    293     lazy = kwds.get('lazy', False)
    294     file_data_list = reader.file_reader(filename,
--> 295                                         **kwds)
    296     objects = []
    297 

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in file_reader(filename, record_by, order, lazy)
    986     with open(filename, "rb") as f:
    987         dm = DigitalMicrographReader(f)
--> 988         dm.parse_file()
    989         images = [ImageObject(imdict, f, order=order, record_by=record_by)
    990                   for imdict in dm.get_image_dictionaries()]

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_file(self)
     85             number_of_root_tags,
     86             group_name="root",
---> 87             group_dict=self.tags_dict)
     88 
     89     def parse_header(self):

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    222             else:
    223                 _logger.debug('File address:', self.f.tell())
--> 224                 raise DM3TagIDError(tag_header['tag_id'])
    225 
    226     def get_data_reader(self, enc_dtype):

DM3TagIDError: 66
```
Lacking of knowledge I'm still trying to troubleshoot the issue without almost any success... As the file size is > 50GB, I have exported tags as .gtg file which is attached. 

[02_0V settings.gtg.zip](https://github.com/hyperspy/hyperspy/files/2282667/02_0V.settings.gtg.zip)

**Comment by woozey** [comment]
Here is the part of the logs before it breaks:
``` python
DEBUG:hyperspy.io_plugins.digital_micrograph:Reading tag name at address: 4227702225
DEBUG:hyperspy.io_plugins.digital_micrograph:Tag name: BH;8BFBFGF
DEBUG:hyperspy.io_plugins.digital_micrograph:Tag ID: 66
```

As an intermediate solution, I'll delete the tag using DM and load it again. Meanwhile, I will try to find out what are the tags with ID 66...

**Comment by woozey** [comment]
Failed finding the tag using DM...

**Comment by magnunor** [comment]
Could you generate a small 4-D DM4 file with the same error?

**Comment by woozey** [comment]
I have tried, but haven't worked... The background is that the data comes from K2-IS used together with so called STEMx. The last one is a synchronisation for scanning, i.e. for doing pixelated STEM with the K2 or OneView cameras. K2 dumps data into binary files, on drives of individual processing units directly. Than there is a conversion step to convert that into RAW, DM3 or DM4 files. The first two though, cannot be used converting STEM data. During the conversion step one can choose binning for any of the axis.

So I have tried to bin the data to very small size during the conversion step, but than there was no issue with loading it with HS...

It is all very puzzling. I'll see how small I can get the files that the error still persist. Everything bellow 30 GB I'll be able to upload to a cloud...

**Comment by magnunor** [comment]
Could you acquire a dataset with a low number of probe positions? For example 4 x 4?

That is how the file I/O of Merlin data is tested on `fpd`: https://gitlab.com/fpdpy/fpd/tree/master/fpd/tests/file_datasets

**Comment by woozey** [comment]
I have such data sets, give me few hours to convert the data.

**Comment by woozey** [comment]
Meanwhile with binning 4 (about 16GB, which I can share) I got a new error ))))
```python
ss = hs.load('D:\\Vadim\\2018-08-02_STEM-d_C_fiber_Fengshan\\bin4\\02_0V.dm4')

---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-10-862092a6742c> in <module>()
----> 1 ss = hs.load('D:\\Vadim\\2018-08-02_STEM-d_C_fiber_Fengshan\\bin4\\02_0V.dm4')

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load(filenames, signal_type, stack, stack_axis, new_axis_name, lazy, **kwds)
    241             objects = [load_single_file(filename, lazy=lazy,
    242                                         **kwds)
--> 243                        for filename in filenames]
    244 
    245         if len(objects) == 1:

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in <listcomp>(.0)
    241             objects = [load_single_file(filename, lazy=lazy,
    242                                         **kwds)
--> 243                        for filename in filenames]
    244 
    245         if len(objects) == 1:

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load_single_file(filename, signal_type, **kwds)
    284                                 reader=reader,
    285                                 signal_type=signal_type,
--> 286                                 **kwds)
    287 
    288 

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io.py in load_with_reader(filename, reader, signal_type, **kwds)
    293     lazy = kwds.get('lazy', False)
    294     file_data_list = reader.file_reader(filename,
--> 295                                         **kwds)
    296     objects = []
    297 

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in file_reader(filename, record_by, order, lazy)
    986     with open(filename, "rb") as f:
    987         dm = DigitalMicrographReader(f)
--> 988         dm.parse_file()
    989         images = [ImageObject(imdict, f, order=order, record_by=record_by)
    990                   for imdict in dm.get_image_dictionaries()]

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_file(self)
     85             number_of_root_tags,
     86             group_name="root",
---> 87             group_dict=self.tags_dict)
     88 
     89     def parse_header(self):

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    219                     ntags=ntags,
    220                     group_name=tag_name,
--> 221                     group_dict=group_dict[tag_name])
    222             else:
    223                 _logger.debug('File address:', self.f.tell())

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in parse_tags(self, ntags, group_name, group_dict)
    154                         raise IOError("Expected 20 (string), got %i" % enctype)
    155                     size, enc_eltype = self.parse_array_definition()
--> 156                     data = self.read_array(size, enc_eltype, skip=skip)
    157                 elif infoarray_size > 3:
    158                     enctype = iou.read_long(self.f, "big")

C:\Users\migunov\AppData\Local\conda\conda\envs\hs\lib\site-packages\hyperspy\io_plugins\digital_micrograph.py in read_array(self, size, enc_eltype, extra, skip)
    381                         "size_bytes": size_bytes,
    382                         "offset": self.f.tell()}
--> 383                 self.f.seek(size_bytes, 1)  # Skipping data
    384             else:
    385                 data = eltype(skip=skip, **extra)

OSError: [Errno 22] Invalid argument
```
Any ideas? Shall I start uploading this one?

**Comment by woozey** [comment]
I can't reproduce the original error on smaller datasets. The one above appears quite often though. It seems to me that it is related to the file size...

**Comment by francisco-dlp** [comment]
I was waiting for this bug to hit! The issue is probably that, when I adapted the dm3 for dm4 files, every-time that  I skipped 4 bytes (e.g. [here](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/io_plugins/digital_micrograph.py#L263)) I should have read an int64 instead of an int32. At the time (and still) I didn't have any file big enough to test if that was the case.

**Comment by francisco-dlp** [comment]
@woozey, I've just pull requested a fix #2046, could you test it and let us know if it works? If it works, it'll be nice to merge it for v1.4. Thanks!
